be/src/storage/rowset/beta_rowset_writer.cpp

```diff
@@ -181,13 +191,18 @@ HorizontalBetaRowsetWriter::~HorizontalBetaRowsetWriter() {
                 LOG_IF(WARNING, !st.ok()) << "Fail to delete file=" << tmp_segment_file << ", " << st.to_string();
             }
             _tmp_segment_files.clear();
-            for (auto i = 0; i < _num_segment; ++i) {
-                auto path = BetaRowset::segment_file_path(_context.rowset_path_prefix, _context.rowset_id, i);
-                // Even if an error is encountered, these files that have not been cleaned up
-                // will be cleaned up by the GC background. So here we only print the error
-                // message when we encounter an error.
-                auto st = _fs->delete_file(path);
-                LOG_IF(WARNING, !st.ok()) << "Fail to delete file=" << path << ", " << st.to_string();
+
+            // if final merge hasn't started writing to the new file, the _num_segment
+            // is the number of temp segment file, and we should skip this delete.
+            if (_final_merge_start_write) {
```

Comment:
When will these temporary files be deleted?

Reply:
Temporary files are only used by primary key rowset. For primary key, before finishing, all segemnt data will be written to the temporary files. When we finish all writes, we will call _final_merge to merge all temporary segment files to the new segment files, and then we delete these  temporary files.